src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/RestoreCommandHandler.cs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.ServiceHub.Framework;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Shell.ServiceBroker;
+using Microsoft.VisualStudio.Threading;
+using NuGet.VisualStudio.Internal.Contracts;
+using IAsyncServiceProvider = Microsoft.VisualStudio.Shell.IAsyncServiceProvider;
+using Task = System.Threading.Tasks.Task;
+
+namespace NuGet.VisualStudio.OnlineEnvironment.Client
+{
+    internal class RestoreCommandHandler
+    {
+        private readonly JoinableTaskFactory _joinableTaskFactory;
+        private readonly IAsyncServiceProvider _asyncServiceProvider;
+        private CancellationTokenSource _cancelBuildToken;
+
+        public RestoreCommandHandler(JoinableTaskFactory joinableTaskFactory, IAsyncServiceProvider asyncServiceProvider)
+        {
+            _joinableTaskFactory = joinableTaskFactory ?? throw new ArgumentNullException(nameof(joinableTaskFactory));
+            _asyncServiceProvider = asyncServiceProvider ?? throw new ArgumentNullException(nameof(asyncServiceProvider));
+        }
+
+        public void RunSolutionRestore()
+        {
+            _joinableTaskFactory.RunAsync(RunSolutionRestoreAsync);
```

Comment:
`RunAsync` returns `JoinableTask`. In this case we are not storing the return value to `await` later.

Reply:
I don't get what you are suggesting?
It's not awaited on purpose, the method is file and forget. If we block, we block the UI.

Reply:
> It's not awaited on purpose, the method is file and forget. If we block, we block the UI

Thanks for the clarification. Looks good because the method is file and forget.